commit/plugin.go

```diff
@@ -85,8 +102,147 @@ func NewPlugin(
+	if err != nil {
+		//log error
+		p.lggr.Errorw("failed to get token prices", "err", err)
+	}
+	gasObs, err := p.feeProcessor.Observation(ctx, prevOutcome.ChainFeeOutcome, chainfee.Query{})
+	if err != nil {
+		p.lggr.Errorw("failed to get gas prices", "err", err)
 	}
+
+	obs := Observation{
+		MerkleObs: merkleObs,
+		TokenObs:  tokenPricesObs,
+		GasObs:    gasObs,
+		FChain:    fChain,
+	}
+	return obs.Encode()
+}
+
+func (p *Plugin) ObserveFChain() map[cciptypes.ChainSelector]int {
+	fChain, err := p.homeChain.GetFChain()
+	if err != nil {
+		// TODO: metrics
+		p.lggr.Warnw("call to GetFChain failed", "err", err)
+		return map[cciptypes.ChainSelector]int{}
+	}
+	return fChain
+}
+
+// Outcome depending on the current state, either:
+// - chooses the seq num ranges for the next round
+// - builds a report
+// - checks for the transmission of a previous report
+func (p *Plugin) Outcome(
+	outCtx ocr3types.OutcomeContext, query types.Query, aos []types.AttributedObservation,
+) (ocr3types.Outcome, error) {
+	prevOutcome := p.decodeOutcome(outCtx.PreviousOutcome)
+
+	var merkleObservations []MerkleRootObservation
+	var tokensObservations []TokenPricesObservation
+	var feeObservations []ChainFeeObservation
+
+	for _, ao := range aos {
+		obs, err := DecodeCommitPluginObservation(ao.Observation)
+		if err != nil {
+			p.lggr.Errorw("failed to decode observation", "err", err)
+			continue
+		}
+		merkleObservations = append(merkleObservations,
+			MerkleRootObservation{
+				OracleID:    ao.Observer,
+				Observation: obs.MerkleObs,
+			},
+		)
+
+		tokensObservations = append(tokensObservations,
+			TokenPricesObservation{
+				OracleID:    ao.Observer,
+				Observation: obs.TokenObs,
+			},
+		)
+
+		feeObservations = append(feeObservations,
+			ChainFeeObservation{
+				OracleID:    ao.Observer,
+				Observation: obs.GasObs,
+			},
+		)
+	}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(3)
+
+	var merkleOutcome merkleroot.Outcome
+	var tokensOutcome tokenprice.Outcome
+	var gasOutcome chainfee.Outcome
+
+	go func(mOut *merkleroot.Outcome) {
+		defer wg.Done()
+		outcome, err := p.merkleProcessor.Outcome(
+			prevOutcome.MerkleRootOutcome,
+			merkleroot.Query{},
+			merkleObservations,
+		)
+		if err != nil {
+			p.lggr.Errorw("failed to get merkle outcome", "err", err)
+		}
+
+		*mOut = outcome
+	}(&merkleOutcome)
+
+	go func(tOut *tokenprice.Outcome) {
+		defer wg.Done()
+		outcome, err := p.tokenProcessor.Outcome(prevOutcome.TokenPriceOutcome, tokenprice.Query{}, tokensObservations)
+		if err != nil {
+			p.lggr.Errorw("failed to get token prices outcome", "err", err)
+		}
+		*tOut = outcome
+	}(&tokensOutcome)
+
+	go func(gOut *chainfee.Outcome) {
+		defer wg.Done()
+		outcome, err := p.feeProcessor.Outcome(prevOutcome.ChainFeeOutcome, chainfee.Query{}, feeObservations)
+		if err != nil {
+			p.lggr.Errorw("failed to get gas prices outcome", "err", err)
+		}
+		*gOut = outcome
+	}(&gasOutcome)
+
+	wg.Wait()
```

Comment:
If you want to insist on using a concurrent approach here lets use errorgroup instead, since it seems like none of the errors above are checked outside of the goroutines themselves.

I wonder if the outcome should be "all or nothing" i.e if one of the processors has an error, should we error out of the outcome completely? The reasoning being that there's no way to know only from the outcome if one of the outcomes is empty due to an error or if its truly empty.